File: sdk/typescript/rooch-sdk/src/transactions/transaction.test.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { describe, it, expect } from 'vitest'

import { bcs } from '../bcs/index.js'
import { str } from '../utils/index.js'
import { BitcoinSignMessage } from '../crypto/index.js'
import { Secp256k1Keypair } from '../keypairs/index.js'
import { Transaction } from '../transactions/index.js'

describe('Transactions', () => {
  it('verify transaction', async () => {
    const signer = new Secp256k1Keypair()

    const tx = new Transaction()
    tx.callFunction({
      target: '0x3::empty::empty_with_signer',
    })

    tx.setSender(signer.getRoochAddress().toHexAddress())
    tx.setSeqNumber(BigInt(0))
    tx.setChainId(BigInt(4))

    const auth = await signer.signTransaction(tx)

    const payload = bcs.BitcoinAuthPayload.parse(auth.payload)

    const bitcoinMessage = new BitcoinSignMessage(tx.hashData(), str('utf8', payload.messageInfo))

    const result = await signer.getPublicKey().verify(bitcoinMessage.hash(), payload.signature)
    expect(result).toBeTruthy()
  })
})

File: sdk/golang/rooch-sdk/transactions/transaction_test.go
package transactions

import (
	"testing"

	"github.com/rooch-network/rooch/sdk/golang/bcs"
	"github.com/rooch-network/rooch/sdk/golang/crypto"
	"github.com/rooch-network/rooch/sdk/golang/keypairs"
	"github.com/stretchr/testify/assert"
)

func TestTransactionVerification(t *testing.T) {
	// Create a new keypair
	signer, err := keypairs.NewSecp256k1Keypair()
	assert.NoError(t, err)

	// Create and configure transaction
	tx := NewTransaction()
	tx.CallFunction(&FunctionCall{
		Target: "0x3::empty::empty_with_signer",
	})

	tx.SetSender(signer.GetRoochAddress().ToHexAddress())
	tx.SetSeqNumber(0)
	tx.SetChainId(4)

	// Sign the transaction
	auth, err := signer.SignTransaction(tx)
	assert.NoError(t, err)

	// Parse the auth payload
	payload, err := bcs.ParseBitcoinAuthPayload(auth.Payload)
	assert.NoError(t, err)

	// Create bitcoin message
	bitcoinMessage := crypto.NewBitcoinSignMessage(tx.HashData(), []byte(payload.MessageInfo))

	// Verify signature
	result, err := signer.GetPublicKey().Verify(bitcoinMessage.Hash(), payload.Signature)
	assert.NoError(t, err)
	assert.True(t, result)
} 